sdk/src/nonce_state.rs

```diff
@@ -0,0 +1,545 @@
+    Uninitialized,
+    Initialized(Meta, Hash),
+}
+
+impl Default for NonceState {
+    fn default() -> Self {
+        Self::Uninitialized
+    }
+}
+
+impl NonceState {
+    pub fn size() -> usize {
+        bincode::serialized_size(
+            &NonceState::Initialized(
+                Meta::default(),
+                Hash::default(),
+        )).unwrap() as usize
+    }
+}
+
+pub trait NonceAccount {
+    fn nonce(
+        &mut self,
+        recent_blockhashes: &RecentBlockhashes,
+        rent: &Rent,
+    ) -> Result<(), InstructionError>;
+    fn set_nonce(
+        &mut self,
+        recent_blockhashes: &RecentBlockhashes,
+        meta: &Meta,
+    ) -> Result<(), InstructionError>;
+    fn withdraw(
+        &mut self,
+        lamports: u64,
+        to: &mut KeyedAccount,
+        recent_blockhashes: &RecentBlockhashes,
+        signers: &HashSet<Pubkey>,
+    ) -> Result<(), InstructionError>;
+}
+
+impl<'a> NonceAccount for KeyedAccount<'a> {
+    fn set_nonce(
+        &mut self,
+        recent_blockhashes: &RecentBlockhashes,
+        meta: &Meta,
+    ) -> Result<(), InstructionError> {
+        if recent_blockhashes.is_empty() {
+            Err(NonceError::NoRecentBlockhashes.into())
+        } else {
+            self.set_state(
+                &NonceState::Initialized(*meta, recent_blockhashes[0])
+            )
+        }
+    }
+
+    fn nonce(
+        &mut self,
+        recent_blockhashes: &RecentBlockhashes,
+        rent: &Rent,
+    ) -> Result<(), InstructionError> {
+        let meta = match self.state()? {
+            NonceState::Initialized(meta, ref hash) => {
+                if *hash == recent_blockhashes[0] {
+                    return Err(NonceError::NotExpired.into());
+                }
+                meta
+            },
+            NonceState::Uninitialized => {
+                let min_balance = rent.minimum_balance(self.account.data.len());
+                if self.account.lamports < min_balance {
+                    return Err(InstructionError::InsufficientFunds.into());
+                }
+                Meta::new(min_balance)
+            },
+        };
+        self.set_nonce(recent_blockhashes, &meta)
+    }
+
+    fn withdraw(
+        &mut self,
+        lamports: u64,
+        to: &mut KeyedAccount,
+        recent_blockhashes: &RecentBlockhashes,
+        signers: &HashSet<Pubkey>,
+    ) -> Result<(), InstructionError> {
+        if !signers.contains(self.unsigned_key()) {
+            return Err(InstructionError::MissingRequiredSignature);
+        }
+
+        match self.state()? {
+            NonceState::Uninitialized => {
+                if lamports > self.account.lamports {
+                    return Err(InstructionError::InsufficientFunds.into());
+                }
+            },
+            NonceState::Initialized(meta, ref hash) => {
+                if lamports == self.account.lamports {
+                    if *hash == recent_blockhashes[0] {
```

Comment:
Wouldn't this be a way to cancel it?  Do we not want to nonce cancellation?

Reply:
We must ensure that the nonce is replaced with a new value to prevent replay.

In what context are you referring to cancellation here?

Reply:
Replay?  I guess I'm confused about why we allow withdrawing in the first place if we can't withdraw all the lamports are we just providing a mechanism because someone might deposit too many lamports during account creation?

I would also think we would want a way to cancel a nonce account and extract all the lamports once its no longer needed.  When a nonce account is canceled it just means that lamport balance goes to zero and any further reference to it will fail

Reply:
If the stored nonce value does not advance, it would allow a transaction using this featured to be resubmitted to the network (replayed) and succeed again.

Thanks for clarifying "cancellation", it's what I've been calling "close" in my head.  The `Withdraw` stuff came in with Rent Exemption, which I don't fully grok the minutia on yet.  I pretty much monkey-see-monkey-do'd it out of staking to get things moving.  I agree that an instruction that just does the full withdraw and allows the account to be cleaned up seems more ergonomic, but I'm not sure of the constraints of closing a rent exempt account.

Reply:
I guess the question is also do we want to give the nonce owner the ability to invalidate a transaction they've signed.  Seems like they already have this ability inadvertently by signing a transaction and submitting it before the first.  The first would then have a non-matching nonce and be rejected.